docs/design-documents/20221028-dead-letter-queue.md

```diff
@@ -0,0 +1,192 @@
+
+The purpose of a DLQ is to make a pipeline resilient to errors that might
+happen while processing messages.
+
+The design for the DLQ should support the following:
+
+- Allow user to reroute nacked messages to any Conduit connector (builtin or
+  standalone).
+- A message sent to the DLQ should reflect the contents it had just before the
+  error happened. Example: if a message went through 2 processors and failed
+  while being processed by a 3rd processor, the DLQ message should reflect any
+  transforms done by the first 2 processors.
+- Ensure message order is preserved when messages are sent to the DLQ.
+  In other words - the order in which messages were produced by the source
+  connector should be preserved, even if messages are nacked in a different
+  order (this is already assured using
+  a [semaphore](https://github.com/ConduitIO/conduit/blob/b6325584d51ec64bc9086faddcdaa2788a5dfa8f/pkg/pipeline/stream/source_acker.go#L35-L37)).
+- If a message failed to be sent to the DLQ the nack action should fail and
+  stop the pipeline.
+- Allow the user to configure the window and number of messages that need to be
+  nacked to consider the pipeline broken and trigger a stop. Example: if 25 of
+  the last 30 messages are nacked stop the pipeline.
+- Messages pushed to the DLQ should contain metadata describing the error and
+  the source connector.
+
+### Questions
+
+- **Should we allow the user to attach custom metadata to messages pushed to
+  the DLQ?**
+
+### Background
+
+Currently Conduit pipelines fail fast. If a message fails to be processed for
+any reason it is negatively acknowledged which causes the pipeline to stop
+running as soon as possible. In reality it's often the case that we don't want
+to stop the pipeline because of one bad message, instead we want to store it
+somewhere safe and keep the pipeline running. The solution is the support for
+dead letter queues or DLQs.
+
+## Implementation
+
+### DLQ Handlers
+
+Internally we already have a place to implement DLQ handlers
+([here](https://github.com/ConduitIO/conduit/blob/d19379efc04d20d12ab9c80df82a29fcef7e8afd/pkg/pipeline/stream/source_acker.go#L114-L118)).
+We need to decide on the API of the DLQ handler, in this design we propose an
+interface with a method that takes a context, a record and an error and returns
+an error. This should be all we need to write that record to the DLQ.
+
+```go
+type DLQHandler struct {
+    func Write(ctx context.Context, r record.Record, reason error) error
+}
+```
+
+The function `DLQHandler.Write` is expected to synchronously write the record
+to the DLQ and return `nil` if that action succeeded or an error otherwise. The
+synchronous write is needed to ensure the DLQ message actually reached its
+destination, otherwise we need to consider the nack failed and need to stop the
+pipeline.
+
+### DLQ Strategies
+
+We propose to implement 2 DLQ strategies:
+
+- `log` - This strategy would cause a nacked message to be logged to the
```

Comment:
Maybe we add a logging connector, and reduce this to one strategy?

Reply:
I like that 👍 I'll try changing the config to see how it would turn out if we do this.

Reply:
Thanks for the suggestion, I think this simplified the config a lot. WDYT?